### PR TITLE
expression: don't propagateColumnEQ joinCondition when nullSensitive

### PR DESCRIPTION
**Pull request by wshwsh12**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/23889 

Problem Summary:

### What is changed and how it works?

Proposal: [xxx](url) 

What's Changed:

How it Works:

### Related changes

- Need to cherry-pick to the release branch 4.0, 5.0

### Check List 

Tests 

- Unit test
- Integration test

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

- Fix wrong results for in clause.

### CODE REVIEW

**Review by lzmhhh123 — APPROVED** [review]

File: expression/constant_propagation.go [hunk 1]
```diff
@@ -576,7 +576,9 @@ func (s *propOuterJoinConstSolver) propagateColumnEQ() {
 			if outerCol == nil {
 				continue
 			}
-			visited = s.deriveConds(outerCol, innerCol, mergedSchema, lenJoinConds, visited, false)
+			if !s.nullSensitive {
+				visited = s.deriveConds(outerCol, innerCol, mergedSchema, lenJoinConds, visited, false)
+			}
```
Comment (eurekaka) [inline expression/constant_propagation.go:581]: Seems like we should include line 582 into this check as well?

**Review by eurekaka — APPROVED** [review]
/lgtm

**Pull request merged**